File: publication-figure-renderer/python/src/publication_figure_renderer/svg_deidentifier.py
"""Pure transform: SVG text + {old_name -> new_label} -> SVG text.

``lxml.etree`` parses the SVG into a proper node tree so we never touch
attribute strings, never compete with XML escaping rules, and cannot be
fooled by adversarial characters in display names.
"""

from __future__ import annotations

from collections.abc import Mapping

from lxml import etree

_SVG_NS = "http://www.w3.org/2000/svg"
_TEXT_QNAME = f"{{{_SVG_NS}}}text"


class InvalidSvgError(ValueError):
    """Raised when the input cannot be parsed as XML/SVG."""


def deidentify_svg(
    svg_text: str,
    name_to_label: Mapping[str, str],
    *,
    width: int | None = None,
    height: int | None = None,
) -> str:
    """Return the SVG with display names replaced and optional size overrides.

    An empty-string label removes the matching ``<text>`` element entirely,
    so a ``--label-style=off`` pass still produces valid SVG.
    """
    try:
        root = etree.fromstring(svg_text.encode("utf-8"))
    except etree.XMLSyntaxError as exc:
        raise InvalidSvgError(f"Could not parse SVG: {exc}") from exc

    _apply_dimensions(root, width, height)
    _apply_label_replacements(root, name_to_label)

    return etree.tostring(root, encoding="unicode")


def _apply_dimensions(
    root: etree._Element,
    width: int | None,
    height: int | None,
) -> None:
    if width is not None:
        root.set("width", str(width))
    if height is not None:
        root.set("height", str(height))


def _apply_label_replacements(
    root: etree._Element,
    name_to_label: Mapping[str, str],
) -> None:
    if not name_to_label:
        return
    # iter() with a namespaced qname is both typed cleanly and avoids any
    # xpath-string quoting concerns around adversarial input.
    for node in list(root.iter(_TEXT_QNAME)):
        current = _full_text(node)
        if current not in name_to_label:
            continue
        replacement = name_to_label[current]
        if replacement == "":
            parent = node.getparent()
            if parent is not None:
                parent.remove(node)
        else:
            _set_text(node, replacement)


def _full_text(node: etree._Element) -> str:
    """Return the text content of a <text> node, including any <tspan> children."""
    parts: list[str] = []
    if node.text is not None:
        parts.append(node.text)
    for child in node:
        parts.append(_full_text(child))
        if child.tail is not None:
            parts.append(child.tail)
    return "".join(parts)


def _set_text(node: etree._Element, replacement: str) -> None:
    # Clear any child <tspan> nodes so the replacement does not concatenate
    # with stale text.  lxml escapes the replacement automatically on
    # serialisation -- we never build XML by string concatenation.
    for child in list(node):
        node.remove(child)
    node.text = replacement

File: publication-figure-renderer/r/R/svg_deidentifier.R
# Pure transform: (SVG text + old_name -> new_label mapping) -> SVG text
# with display names replaced inside <text> elements only.
#
# xml2 parses the document into a proper node tree, so we never touch
# attribute strings, never compete with SVG's CDATA/escaping rules, and
# cannot be fooled by adversarial characters in names.

SVG_NS <- "http://www.w3.org/2000/svg"

#' Raise an invalid_svg_error condition.
#' @noRd
stop_invalid_svg <- function(message) {
  stop(structure(
    list(message = message, call = sys.call(-1)),
    class = c("invalid_svg_error", "error", "condition")
  ))
}

#' Return the input SVG text with display names replaced and optional
#' width/height overrides applied to the root `<svg>` element.
#'
#' @param svg_text Raw SVG document as a single character string.
#' @param name_to_label Named character vector: names are the original
#'   display names to replace; values are the replacement labels.
#'   An empty-string label removes the matching <text> element entirely,
#'   so a "--label-style=off" pass still leaves valid SVG.
#' @param width Optional integer; overrides the root `width` attribute.
#' @param height Optional integer; overrides the root `height` attribute.
#' @return The transformed SVG as a single character string.
deidentify_svg <- function(svg_text,
                           name_to_label = character(0),
                           width = NULL, height = NULL) {
  doc <- tryCatch(
    xml2::read_xml(svg_text),
    error = function(e) {
      stop_invalid_svg(sprintf("Could not parse SVG: %s", conditionMessage(e)))
    }
  )

  apply_dimensions(doc, width, height)
  apply_label_replacements(doc, name_to_label)

  as.character(doc)
}

apply_dimensions <- function(doc, width, height) {
  root <- xml2::xml_root(doc)
  if (!is.null(width)) {
    xml2::xml_set_attr(root, "width", as.character(width))
  }
  if (!is.null(height)) {
    xml2::xml_set_attr(root, "height", as.character(height))
  }
}

apply_label_replacements <- function(doc, name_to_label) {
  if (length(name_to_label) == 0L) {
    return(invisible(NULL))
  }
  text_nodes <- xml2::xml_find_all(
    doc, "//svg:text",
    ns = c(svg = SVG_NS)
  )
  lookup_names <- names(name_to_label)
  for (node in text_nodes) {
    current <- xml2::xml_text(node)
    if (!current %in% lookup_names) {
      next
    }
    replacement <- unname(name_to_label[[current]])
    if (!nzchar(replacement)) {
      xml2::xml_remove(node)
    } else {
      xml2::xml_text(node) <- replacement
    }
  }
  invisible(NULL)
}
